ls1intum/Artemis
# `Programming exercises`: Generate code hints from solution entries

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
These changes create the code hints from the `ProgrammingExerciseSolutionEntries` and display these in the client.

### Description
A new button has been added that invokes the generation of code hints from solution entries. Currently, only solution entries for structural test cases have been implemented. In #4955, the creation of solution entries for behavioral test cases is implemented. This however has no effect on the implementation of this PR.

In a follow-up, the following points will be covered:
- Merge the SolutionEntries into one entry per file (not sure if handled client- or server-side)
- Implement mechanisms to decide on when a code hint (for a certain task) will be displayed to participating students. The current idea is to display the latest hint for an unsolved task if the result for this task (in particular the related test cases) has not changed for a defined number of participations (in context these are pushes, not commits).
- Implement mechanisms to restrict access to the code hints for students. This is required to prevent students from scraping for the solution.


### Steps for Testing
Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new Java programming exercise 
4. Wait for the solution build plan to finish
5. Click the button `Create Structural Solution Entries`
4. Click the button `Create Behavioral Solution Entries` if #4955 was merged already
6. Click the button `Generate Code Hints` and wait for the server response
7. Click the button `Manage hints`
8. Click on `View` for a Code Hint and make sure that the entries are displayed correctly
9. Click on `Edit` for a Code Hint and make sure you can edit the title, save it and the list updates correspondingly
10. Click on `Delete` for a Code Hint and make sure that the hint is deleted
11. Insert a Code Hint in the problem statement after a task in the markdown and save the exercise.
12. Check that when clicking the `?`-Symbol in the problem statement, the code is displayed correctly.
13. Check that the previous steps are also possible for text hints.
14. Import the exercise into another course and check that all hints are still there.

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| CodeHintService.java | 83% | 89.64% |
| CodeHintResource.java | 50% | 92.3% |
| ExerciseHintService.java | 75% | 100% |
| code-hint-container.component.ts | 77.77% | 100% |
| solution-entry.component.ts | 66.66% | 100% |

### Screenshots
"Manage Hints":
![Manage Hints](https://user-images.githubusercontent.com/73833780/165367441-d02bee8b-6807-4edc-a0a1-ae17b138bb42.png)

Display a Code Hint:
![Display a Code Hint](https://user-images.githubusercontent.com/73833780/165367521-75722ee3-34dc-4fd2-9705-38c653ddb837.png)

Edit a Code Hints' Title:
![Edit Title](https://user-images.githubusercontent.com/73833780/165367602-ad29f737-92f6-4c73-8175-038f73e9cba1.png)

Display a Code Hint from the Problem Statement:
![Display from Problem Statement](https://user-images.githubusercontent.com/73833780/165367675-360da4b8-71e6-4e70-af4e-960216afcda9.png)
